x/feedist/handler.go

```diff
@@ -0,0 +1,300 @@
+	// previously selected recepients ever being paid.
+	if err := distribute(db, h.ctrl, RevenueAccount(msg.RevenueID), rev.Recipients); err != nil {
+		return res, errors.Wrap(err, "cannot distribute")
+	}
+
+	rev.Recipients = msg.Recipients
+	obj := orm.NewSimpleObj(msg.RevenueID, rev)
+	if err := h.bucket.Save(db, obj); err != nil {
+		return res, errors.Wrap(err, "cannot save")
+	}
+	return res, nil
+}
+
+func (h *updateRevenueHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*UpdateRevenueMsg, error) {
+	rmsg, err := tx.GetMsg()
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot get message")
+	}
+	msg, ok := rmsg.(*UpdateRevenueMsg)
+	if !ok {
+		return nil, errors.InvalidMsgErr.New("unknown transaction type")
+	}
+	if err := msg.Validate(); err != nil {
+		return msg, err
+	}
+	return msg, nil
+}
+
+// distribute split the funds stored under the revenue address and distribute
+// them according to recipients proportions. When successful, revenue account
+// has no funds left after this call.
+//
+// It might be that not all funds can be distributed equally. Because of that a
+// small leftover can remain on the revenue account after this operation.
+func distribute(db weave.KVStore, ctrl CashController, source weave.Address, recipients []*Recipient) error {
+	var chunks int64
+	for _, r := range recipients {
+		chunks += int64(r.Weight)
+	}
+
+	// Find the greatest common division for all weights. This is needed to
+	// avoid leaving big fund leftovers on the source account when
+	// distributing between many recipients. Or when there is only one
+	// recipient with a high weight value.
+	var weights []int32
+	for _, r := range recipients {
+		weights = append(weights, r.Weight)
+	}
+	div := findGcd(weights...)
+
+	chunks = chunks / int64(div)
+
+	balance, err := ctrl.Balance(db, source)
+	switch {
+	case err == nil:
+		// All good.
+	case errors.Is(errors.NotFoundErr, err):
+		// Account does not exist, so there is are no funds to split.
+		return nil
+	default:
+		return errors.Wrap(err, "cannot acquire revenue account balance")
+	}
+
+	// TODO normalize balance. There is no functionality that allows to
+	// normalize x.Coins right now (14 Feb 2019).
+
+	// For each currency, distribute the coins equally to the weight of
+	// each recipient. This can leave small amount of coins on the original
+	// account.
+	for _, c := range balance {
+		// Ignore those coins that have a negative value. This
+		// functionality is supposed to be distributing value from
+		// revenue account, not collect it. Otherwise this could be
+		// used to charge the recipients instead of paying them.
+		if !c.IsPositive() {
+			continue
+		}
+
+		// Rest of the division can be ignored, because we transfer
+		// funds to each recipients separately. Any leftover will be
+		// left on the recipients account.
+		one, _ := c.Divide(chunks)
+
+		for _, r := range recipients {
+			amount := one.Multiply(int64(r.Weight / div))
```

Comment:
req: the distribution code can be simplified and gives better results by dropping the GCD and working with the raw values: `amount, rest := c.Multiply(int64(r.Weight)).Divide(chunks)`. Please see #322

Reply:
Mutliply then divide.... Yeah, that works... I think Piotr was worried about overflow, but in the end that should not be a problem, since we use int64 and whole values should never need more than around 32 bits, so we have 31 bits to play with